dask/array/tiledb_io.py

```diff
@@ -0,0 +1,142 @@
+from . import core
+
+
+def _tiledb_to_chunks(tiledb_array):
+    schema = tiledb_array.schema
+    return list(schema.domain.dim(i).tile for i in range(schema.ndim))
+
+
+def from_tiledb(uri, attribute=None, chunks=None,
+                storage_options=None, **kwargs):
+    """Load array from the TileDB storage format
+
+    See https://docs.tiledb.io for more information about TileDB.
+
+    Parameters
+    ----------
+    uri: TileDB array or str
+        Location to save the data
+    attribute: str or None
+        Attribute selection (single-attribute view on multi-attribute array)
+
+    Examples
+    --------
+
+    >>> # create a tiledb array
+    >>> import tiledb, numpy as np, tempfile  # doctest: +SKIP
+    >>> uri = tempfile.NamedTemporaryFile().name  # doctest: +SKIP
+    >>> tiledb.from_numpy(uri, np.arange(0,9).reshape(3,3))  # doctest: +SKIP
+    <tiledb.libtiledb.DenseArray object at 0x...>
+    >>> # read back the array
+    >>> import dask.array as da  # doctest: +SKIP
+    >>> tdb_ar = da.from_tiledb(uri)  # doctest: +SKIP
+    >>> tdb_ar.shape  # doctest: +SKIP
+    (3, 3)
+    >>> tdb_ar.mean().compute()  # doctest: +SKIP
+    4.0
+    """
+    import tiledb
+    tiledb_config = storage_options or dict()
+    key = tiledb_config.pop('key', None)
+
+    if isinstance(uri, tiledb.Array):
+        tdb = uri
+    else:
+        tdb = tiledb.open(uri, attr=attribute, config=tiledb_config, key=key)
+
+    if tdb.schema.sparse:
+        raise ValueError("Sparse TileDB arrays are not supported")
+
+    if not attribute:
+        if tdb.schema.nattr > 1:
+            raise TypeError("keyword 'attribute' must be provided"
+                            "when loading a multi-attribute TileDB array")
+        else:
+            attribute = tdb.schema.attr(0).name
+
+    if tdb.iswritable:
+        raise ValueError("TileDB array must be open for reading")
+
+    chunks = chunks or _tiledb_to_chunks(tdb)
+
+    assert(len(chunks) == tdb.schema.ndim)
+
+    return core.from_array(tdb, chunks, name='tiledb-%s' % uri)
+
+
+def to_tiledb(darray, uri, compute=True, return_stored=False,
+              storage_options=None, **kwargs):
+    """Save array to the TileDB storage format
+
+    Save 'array' using the TileDB storage manager, to any TileDB-supported URI,
+    including local disk, S3, or HDFS.
+
+    See https://docs.tiledb.io for more information about TileDB.
+
+    Parameters
+    ----------
+
+    darray: dask.array
+        A dask array to write.
+    uri:
+        Any supported TileDB storage location.
+    storage_options:
+        Dict containing any configuration options for the TileDB backend.
+        see https://docs.tiledb.io/en/stable/tutorials/config.html
+    compute, return_stored: see ``store()``
+    :return:
```

Comment:
The syntax here I suspect is wrong for numpydoc - just omit it if there's no return value anyway.
Some storage functions allow a compute=False option to return unevaluated delayed objects, you might want to add that (perhaps in another PR).